### PR TITLE
[Release/202405] Moves placement of variable policy registration to after the last variable that is "Read-Only"

### PR DESCRIPTION
**Pull request by Flickdm**
## Description

Attach to https://github.com/microsoft/mu_basecore/pull/978

This change moves the locking of the variable policy to below the last Read-Only variable. This is necessary to ensure that the variable policy is not locked until after read only variables are set. Otherwise, the SetVariable On a Read-Only variable will fail with "WRITE_PROTECTED" error.

- [X] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

QC Surface Platform

## Integration Instructions

N/A

### CODE REVIEW

**Review by kenlautner — APPROVED** [review]

File: MdeModulePkg/Universal/BdsDxe/BdsEntry.c [hunk 1]
```diff
@@ -832,6 +815,34 @@ BdsEntry (
...
+
+  //
+  // READ ONLY Variable (registered in mReadOnlyVariables) MUST BE be set before this point.
+  //
+
+  //
+  // Mark the read-only variables if the Variable Lock protocol exists
+  //
+  Status = gBS->LocateProtocol (&gEdkiiVariablePolicyProtocolGuid, NULL, (VOID **)&VariablePolicy);
+  DEBUG ((DEBUG_INFO, "[BdsDxe] Locate Variable Policy protocol - %r\n", Status));
```
Comment (os-d) [inline MdeModulePkg/Universal/BdsDxe/BdsEntry.c]: We can probably just print if we fail to find the protocol.

File: MdeModulePkg/Universal/BdsDxe/BdsEntry.c [hunk 2]
```diff
@@ -832,6 +815,34 @@ BdsEntry (
...
+                 VariablePolicy,
+                 &gEfiGlobalVariableGuid,
+                 mReadOnlyVariables[Index],
+                 VARIABLE_POLICY_NO_MIN_SIZE,
+                 VARIABLE_POLICY_NO_MAX_SIZE,
+                 VARIABLE_POLICY_NO_MUST_ATTR,
+                 VARIABLE_POLICY_NO_CANT_ATTR,
+                 VARIABLE_POLICY_TYPE_LOCK_NOW
+                 );
+      ASSERT_EFI_ERROR (Status);
```
Comment (os-d) [inline MdeModulePkg/Universal/BdsDxe/BdsEntry.c]: I think worth adding a print if we fail to register the policy.

**Review by os-d — APPROVED** [review]
The PR description says we do this to ensure policy is locked before any variable is accessed. Can you change this to indicate we are locking after we finishing setting the variables?

File: MdeModulePkg/Universal/BdsDxe/BdsEntry.c (hunk 1 above)
Comment (makubacki) [reply MdeModulePkg/Universal/BdsDxe/BdsEntry.c]: Probably should even assert if you fail to find the protocol.

**Review by makubacki — APPROVED** [review]

File: MdeModulePkg/Universal/BdsDxe/BdsEntry.c (hunk 1 above)
Comment (Flickdm) [reply MdeModulePkg/Universal/BdsDxe/BdsEntry.c]: Do we expect every edk2 based platform to produce the protocol at this point?

Comment (makubacki) [reply MdeModulePkg/Universal/BdsDxe/BdsEntry.c]: I'd expect every Mu project to not every edk2 platform.

**Pull request merged**